src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Riok.Mapperly.Helpers;
+
+namespace Riok.Mapperly.Descriptors.Enumerables.EnsureCapacity;
+
+public static class EnsureCapacityBuilder
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+    private const string CountMethodName = nameof(ICollection<object>.Count);
+    private const string LengthMethodName = nameof(Array.Length);
+    private const string TryGetNonEnumeratedCountMethodName = "TryGetNonEnumeratedCount";
+
+    public static EnsureCapacity? TryCreateEnsureCapacityBuilder(ITypeSymbol sourceType, ITypeSymbol targetType, WellKnownTypes types)
+    {
+        var capacityMethod = targetType.GetMembers(EnsureCapacityName)
+                        .OfType<IMethodSymbol>()
+                        .FirstOrDefault(x => x.Parameters.Length == 1
+                        && x.Parameters[0].Type.SpecialType == SpecialType.System_Int32
+                        && !x.IsStatic);
+
+        // if EnsureCapacity is not available then return null
+        if (capacityMethod is null)
+            return null;
+
+        // if target does not have a count then return null
+        if (!TryGetNonEnumeratedCount(targetType, types, out var targetSizeProperty))
+            return null;
+
+        // if target and source count are known then create a simple EnsureCapacity statement
+        if (TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+            return new EnsureCapacityMember(targetSizeProperty, sourceSizeProperty);
+
+        sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable);
```

Comment:
Instead of asserting `iEnumerable` is non-null later, I'd if + return null here:
```c#
if (!sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable))
    return null;
```